### PR TITLE
Add Torch.Typed.Optim.CppOptim and associated example

### PR DESCRIPTION
**Pull request by RichardWarfield**
A typed version of Torch.Optim.CppOptim.  I have also added a partial reimplementation of the `optimizers-cpp` example under `optimizers-cpp-typed` to demonstrate usage.

### CODE REVIEW

**Comment by tscholak** [comment]
Awesome @RichardWarfield, thank you very much! Looking forward to reviewing it!

**Comment by junjihashimoto** [comment]
LGTM!

**Review by junjihashimoto — DISMISSED** [review]
Thx!

File: hasktorch/src/Torch/Typed/Optim/CppOptim.hs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      initParams'= hmap' ToDependent $ flattenParameters model
+
+instance CppOptimizer SGDOptions where
+  initOptimizer  opt@SGDOptions{..} model = do
+    v <- cast6 LibTorch.sgd sgdLr sgdMomentum sgdDampening sgdWeightDecay sgdNesterov initParams'
+    return $ CppOptimizerState opt v
+    where
+      initParams'= hmap' ToDependent $ flattenParameters model
+
+runStep :: (CppOptimizer option, Parameterized model
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Typed/Optim/CppOptim.hs:130]: This typed runStep is not the same as Torch.Typed.Optim's one.
Can we use type class for runStep?
If there are any problems on design, there is no need to use it.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/Optim/CppOptim.hs:130]: If it doesn't seem to be easy to fix, I think it's okay to merge this pr.

Comment (RichardWarfield) [reply hasktorch/src/Torch/Typed/Optim/CppOptim.hs:130]: The only difference should be the learningRate parameter which I excluded for the reasons discussed in #467.  The learning rate is part of the `options` field of the `optim` argument already.

**Review by tscholak — APPROVED** [review]
I agree with @junjihashimoto.
I see this as an incremental milestone. I think we will figure out better abstractions in the future.
Thank you very much @RichardWarfield!

**Pull request merged**